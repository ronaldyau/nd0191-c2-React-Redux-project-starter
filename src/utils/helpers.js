import React from 'react';
import { useLocation, useNavigate, useParams } from 'react-router-dom';
import avatar from '../static/avatar.png';

export const formatDate = (timestamp) => {
  const date = new Date(timestamp);
  const time = date.toLocaleTimeString('en-US');
  return time.substr(0, 5) + time.slice(-2) + ' | ' + date.toLocaleDateString();
};

export const formatPoll = (question, author, authedUser) => {
  let hasAnswered = false;
  let isAuthor = false;
  let selectedVote = '';

  const {
    id,
    optionOne: {
      votes: [],
      text: optionOneText,
    },
    optionTwo: {
      votes: [],
      text: optionTwoText,
    },
    timestamp,
  } = question;

  if (question.optionOne.votes.includes(authedUser)) {
    selectedVote = question.optionOne.text;
  } else if (question.optionTwo.votes.includes(authedUser)) {
    selectedVote = question.optionTwo.text;
  }
  hasAnswered = !!selectedVote;

  if (author === authedUser) {
    isAuthor = true;
  }

  const optionOneVotes = question.optionOne.votes.length;
  const optionTwoVotes = question.optionTwo.votes.length;
  const totalVotes = optionOneVotes + optionTwoVotes;

  const optionOnePercentage = Math.round((optionOneVotes / totalVotes) * 100);
  const optionTwoPercentage = Math.round((optionTwoVotes / totalVotes) * 100);

  return {
    id,
    timestamp,
    name: author.name,
    avatarURL: author.avatarURL ?? avatar,
    optionOneText,
    optionTwoText,
    hasAnswered,
    selectedVote,
    isAuthor,
    totalVotes,
    optionOnePercentage,
    optionTwoPercentage,
  };
};

export const withRouter = (Component) => {
  const ComponentWithRouterProp = (props) => {
    let location = useLocation();
    let navigate = useNavigate();
    let params = useParams();
    return <Component {...props} router={{ location, navigate, params }} />;
  };

  return ComponentWithRouterProp;
};