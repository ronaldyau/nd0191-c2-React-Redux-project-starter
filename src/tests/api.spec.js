import { _saveQuestion, _saveQuestionAnswer } from '../utils/_DATA';


// For the _DATA.js file, write an async unit test for _saveQuestion 
// to verify that the saved question is returned and all expected fields are populated when correctly formatted data is passed to the function.

describe('api', () => {
    describe('_saveQuestion', () => {
        test('should return the saved question and all expected fields are populated', async () => {
          const question = {
            optionOneText: 'Option A',
            optionTwoText: 'Option B',
            author: 'Some Person',
          };
      
          const response = await _saveQuestion(question);
          expect(response).toMatchSnapshot({
            author: 'Some Person',
            id: expect.any(String),
            optionOne: { text: 'Option A' },
            optionTwo: { text: 'Option B' },
            timestamp: expect.any(Number),
          });
        });
      
        it('should return error when incorrect arguments are received', async () => {
          const question = {
            optionOneText: undefined,
            optionTwoText: undefined,
            author: 'jsmith',
          };
      
          await expect(_saveQuestion(question)).rejects.toMatchSnapshot();
        });
    });
      
    describe('_saveQuestionAnswer', () => {
        it('should return true when correct arguments are received', async () => {
          const authedUser = 'sarahedo';
          const qid = '6ni6ok3ym7mf1p33lnez';
          const answer = 'optionOne';
      
          const response = await _saveQuestionAnswer({ authedUser, qid, answer });
      
          expect(response).toBeTruthy();
        });
      
        it('should return an error if incorrect arguments are received', async () => {
          const authedUser = 'jsmith';
          const qid = undefined;
          const answer = undefined;
      
          await expect(_saveQuestionAnswer({ authedUser, qid, answer })).rejects.toMatchSnapshot();
        });
      });
})


