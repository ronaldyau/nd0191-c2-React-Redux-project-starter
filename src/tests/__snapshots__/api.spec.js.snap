// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api _saveQuestion should return error when incorrect arguments are received 1`] = `"Please provide optionOneText, optionTwoText, and author"`;

exports[`api _saveQuestion should return the saved question and all expected fields are populated 1`] = `
{
  "author": "Some Person",
  "id": Any<String>,
  "optionOne": {
    "text": "Option A",
    "votes": [],
  },
  "optionTwo": {
    "text": "Option B",
    "votes": [],
  },
  "timestamp": Any<Number>,
}
`;

exports[`api _saveQuestionAnswer should return an error if incorrect arguments are received 1`] = `"Please provide authedUser, qid, and answer"`;
